dali/operators/coord/coord_transform.h

```diff
@@ -0,0 +1,307 @@
+
+    const int N = input_shape.num_samples();
+    input_pt_dim_ = 0;
+    for (int i = 0; i < N; i++) {
+      auto sample_shape = input_shape.tensor_shape_span(i);
+      if (volume(sample_shape) == 0)
+        continue;
+      int pt_dim = input_shape.tensor_shape_span(i).back();
+      if (input_pt_dim_ == 0) {
+        input_pt_dim_ = pt_dim;
+      } else {
+        DALI_ENFORCE(pt_dim == input_pt_dim_, make_string("The point dimensions must be the same "
+        "for all input samples. Got: ", input_shape, "."));
+      }
+    }
+    if (input_pt_dim_ == 0)
+      return;  // data is degenerate - empty batch or a batch of empty tensors
+
+    if (HasFusedMT()) {
+      ProcessMatrixArg(ws, "MT", N);
+    } else {
+      ProcessMatrixArg(ws, "M", N);
+      ProcessTranslationArg(ws, "T", N);
+    }
+  }
+
+  bool HasFusedMT() const {
+    return has_mt_ || has_mt_input_;
+  }
+
+  bool HasMatrixRegularArg() const {
+    return has_matrix_ || has_mt_;
+  }
+
+  bool HasMatrixInputArg() const {
+    return has_matrix_input_ || has_mt_input_;
+  }
+
+  void ProcessMatrixArg(const workspace_t<Backend> &ws, const char *name, int N) {
+    bool is_fused = HasFusedMT();
+    int cols = input_pt_dim_ + (is_fused ? 1 : 0);
+    if (is_fused)
+      translation_.clear();
+
+    if (HasMatrixRegularArg()) {
+      mtx_ = spec_.template GetRepeatedArgument<float>(name);
+      if (mtx_.size() == 1) {
+        output_pt_dim_ = input_pt_dim_;
+        mtx_.resize(output_pt_dim_ * input_pt_dim_);
+        FillDiag(mtx_, mtx_[0]);
+        if (is_fused) {
+          translation_.resize(output_pt_dim_, 0);
+        }
+      } else {
+        DALI_ENFORCE(mtx_.size() % cols == 0, make_string("Cannot form a matrix ",
+          mtx_.size(), " elements and ", cols, "columns"));
+        output_pt_dim_ = mtx_.size() / cols;
+        if (is_fused)
+          SplitFusedMT();
+      }
+      Repeat(per_sample_mtx_, mtx_, N);
+      if (is_fused)
+        Repeat(per_sample_translation_, translation_, N);
+    } else if (HasMatrixInputArg()) {
+      const auto &M_inp = ws.ArgumentInput(name);
+      auto M = view<const float>(M_inp);
+      DALI_ENFORCE(is_uniform(M.shape), "Matrices for all samples int the batch must have "
+        "the same shape.");
+      DALI_ENFORCE(M.shape.sample_dim() == 0 || M.shape.sample_dim() == 2, make_string(
+        "The parameter M must be a list of 2D matrices of same size or a list of scalars. Got: ",
+        M.shape));
+      if (M.shape.sample_dim() == 0) {  // we have a list of scalars - put them on diagonals
+        output_pt_dim_ = cols;
+        int mat_size = output_pt_dim_ * cols;
+        per_sample_mtx_.resize(mat_size * N);
+        for (int i = 0; i < N; i++) {
+          FillDiag(make_span(&per_sample_mtx_[i * mat_size], mat_size), M.data[i][0]);
+        }
+        translation_.resize(output_pt_dim_, 0);
+        Repeat(per_sample_translation_, translation_, N);
```

Comment:
Shouldn't that happen only for `if (is_fused)` case?

Reply:
It should - it will be overwritten, but yes.

Reply:
I would add this to make the code clear and not spread the responsibility for filling the translation_.